lib/stackprof/middleware.rb

```diff
@@ -40,11 +41,20 @@ def enabled?(env)
         end
       end
 
-      def save(filename = nil)
+      def save
         if results = StackProf.results
-          FileUtils.mkdir_p(Middleware.path)
-          filename ||= "stackprof-#{results[:mode]}-#{Process.pid}-#{Time.now.to_i}.dump"
-          File.open(File.join(Middleware.path, filename), 'wb') do |f|
+          path = Middleware.path
+          is_directory = path != path.chomp('/')
```

Comment:
This could still break for any users who were passing a directory without a trailing slash, right?

I was thinking you could use File.directory?

Reply:
It turns out that this is trickier than I expected. Below is the reasoning for the latest changes:
- Ruby's `file?` and `directory?` return `true` only is the file/directory already exist
- These are the tests I did:
```
# try one at a time including an empty options hash
options = {
  # path: 'path/to/dir/'
  # path: 'path/to/dir' # 'dir' will be treated as file name here
  # path: 'path/to/dir/my_file.dump'
  path: ''
}

if options.empty?
  input = nil
else
  input = options[:path]
end

options[:path] = 'tmp/' if options[:path].to_s.empty?

path = options[:path]
is_directory = path != path.chomp('/')

if is_directory
  filename = 'auto_generated.dump'
else
  filename = File.basename(path)
  path = File.dirname(path)
end
p "Input: #{input}"
p "Path: #{path}"
p "File: #{filename}"
p "Joined: #{File.join(path, filename)}"
```

Reply:
- also, if the user passes `path:''`, I override this with `tmp/auto_generated.dump` because it wouldn't make sense (_I believe_).

Reply:
Okay makes sense. Thanks!

Reply:
Hmm it looks like tests are breaking because they expect the old behavior.

Reply:
I am going to give it a try. Will let the tests run through GitHub (Don't use `travis-ci` locally).

Reply:
I guess the last failure is unrelated?